p2p/protocol.go

```diff
@@ -0,0 +1,173 @@
+package p2p
+
+import (
+	"errors"
+	"github.com/gogo/protobuf/proto"
+	"github.com/google/uuid"
+	"github.com/spacemeshos/go-spacemesh/crypto"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/pb"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+	"sync"
+	"time"
+)
+
+type Protocol struct {
+	name               string
+	network            Service
+	pendMutex          sync.RWMutex
+	pending            map[crypto.UUID]chan interface{}
+	resHandlers        map[crypto.UUID]func(msg []byte) interface{}
+	msgRequestHandlers map[string]func(msg []byte) []byte
+	ingressChannel     chan service.Message
+}
+
+func NewProtocol(network Service, name string) *Protocol {
+	p := &Protocol{
+		name:               name,
+		pending:            make(map[crypto.UUID]chan interface{}),
+		resHandlers:        make(map[crypto.UUID]func(msg []byte) interface{}),
+		network:            network,
+		ingressChannel:     network.RegisterProtocol(name),
+		msgRequestHandlers: make(map[string]func(msg []byte) []byte),
+	}
+	go p.readLoop()
+	return p
+}
+
+func (p *Protocol) readLoop() {
+	for {
+		msg, ok := <-p.ingressChannel
+		if !ok {
+			// Channel is closed.
+			break
+		}
+
+		//todo add buffer and option to limit number of concurrent goroutines
+
+		go func(msg service.Message) {
+			headers := &pb.MessageWrapper{}
```

Comment:
refactor to a new function call, also, notice the level of indentation in this func, remove as much as possible

Reply:
thanks !